lib/hyacinth/dynamic_fields_loader.rb

```diff
@@ -2,7 +2,57 @@
 
 module Hyacinth
   module DynamicFieldsLoader
-    def self.load_rights_fields!
+    # Loads fields with a serialized representation of fields. If a field is already present that field is updated with
+    # the new attributes given. If there's a problem loading a field an error will be raised and all changes will be
+    # rolled back. Optionally vocabularies can also be created if a dynamic field depends on it.
+    #
+    # @param field_definitions [Hash] serialized hash containing field definitions
+    #   Format should look like the following example. Each hash should contain all the fields required by that model.
+    #      {
+    #         dynamic_field_categories: [
+    #           {
+    #             display_label: 'Descriptive Metadata',
+    #             metadata_form: 'descriptive',
+    #             dynamic_field_groups: [
+    #               {
+    #                 string_key: 'name',
+    #                 display_label: 'Name',
+    #                 dynamic_field_groups: {
+    #                   string_key: 'role',
+    #                   dynamic_fields: [
+    #                     { string_key: 'term', field_type: 'controlled_term' }
+    #                   ]
+    #                 }
+    #               }
+    #             ]
+    #           }
+    #         ]
+    #       }
+    # @param [opts] options to use when generating fields
+    # @option opts [Boolean] :load_vocabularies option to load any vocabularies fields are dependent on
+    def self.load_fields!(field_definitions, opts = {})
+      # If all this fields are not loaded successfully we need to rollback any changes.
+      ActiveRecord::Base.transaction do
+        # Load Item and Asset Rights Fields
+        field_definitions[:dynamic_field_categories].each do |category_config|
+          display_label = category_config.delete(:display_label)
+          child_dynamic_field_groups = category_config.delete(:dynamic_field_groups)
+
+          category = DynamicFieldCategory.find_or_create_by!(display_label: display_label) do |c|
+            c.assign_attributes(category_config)
+          end
+
+          create_dynamic_field_groups!(category, child_dynamic_field_groups, opts)
```

Comment:
It doesn't have to be for this pull request, but how would you feel if we renamed this module `DynamicFieldsCreator` instead of `DynamicFieldsLoader` (and renamed load_* to create_*)?  Reason is, when I hear "load" it makes me think "retrieving and loading something into memory" rather than saving something to permanent storage.

Reply:
I am open about the name but I would rather not use "DynamicFieldCreator" because its not always creating a new field.

Reply:
Great!  I'm not tied to DynamicFieldCreator, but would mostly just prefer to avoid `Loader`.

Do you mean that you don't like the "Creator" part because it only creates fields that don't already exist?

Reply:
Yes.

Reply:
"Creator" doesn't bother me because in conditions when it actually does something, that something is a "create" operation, but I guess if you want to express the conditional nature of the functionality in the class name (rather than calling it "Creator" and maybe just expressing "_if_exist" in a method name), you could call it `DynamicFieldConditionalCreator`?

Reply:
That sounds really clunky. I would honestly just rather keep the original name. I think load is a general term that shouldn't be tied specifically to loading things into memory.

Reply:
All I mean is that load just doesn't sound accurate for what it's doing, since it's creating rather than loading.  Load also doesn't communicate the conditional thing you were concerned about, since it doesn't load things that have already been created.  But if you have a strong preference for load, feel free to keep it.  It was just a minor thing for me.